Formula/firefoxpwa.rb

```diff
@@ -0,0 +1,83 @@
+  license "MPL-2.0"
+  head "https://github.com/filips123/FirefoxPWA.git"
+
+  depends_on "rust" => :build
+
+  def install
+    cd "native"
+
+    # Set the correct version in the source files
+    inreplace "Cargo.toml",
+              "version = \"0.0.0\"",
+              "version = \"#{version}\""
+    inreplace "userchrome/profile/chrome/pwa/chrome.jsm",
+              "DISTRIBUTION_VERSION = '0.0.0'",
+              "DISTRIBUTION_VERSION = '#{version}'"
+
+    # Set the path in the manifest to the Homebrew libexec directory
+    # This is needed so the program respects Homebrew's installation hierarchy
+    # We can use macOS manifest for Linux because the only difference is a default path
+    inreplace "manifests/macos.json",
+              "/usr/local/libexec/firefoxpwa-connector",
+              "#{libexec}/firefoxpwa-connector"
+
+    # Set the system data path to the Homebrew share directory
+    # This is needed so the program respects Homebrew's installation hierarchy
+    ENV["FFPWA_SYSDATA"] = share
+
+    # Build the native part
+    # We cannot use "cargo install" to automatically install it
+    # Because the package contains multiple binary and other files
+    system "cargo", "build", "--release"
+
+    # Install all files
+    bin.install "target/release/firefoxpwa"
+    libexec.install "target/release/firefoxpwa-connector"
+    share.install "manifests/macos.json" => "firefoxpwa.json"
+    share.install "userchrome/"
+    bash_completion.install "target/release/completions/firefoxpwa.bash" => "firefoxpwa"
+    fish_completion.install "target/release/completions/firefoxpwa.fish"
+    zsh_completion.install "target/release/completions/_firefoxpwa"
+  end
+
+  def post_install
+    ohai "You have successfully installed the native part of the FirefoxPWA project"
+    ohai "You should also install the Firefox extension if you haven't already"
+    ohai "Download: https://addons.mozilla.org/firefox/addon/firefoxpwa/"
+  end
+
+  def caveats
+    filename = "firefoxpwa.json"
+
+    source = share
+    destination = ""
+
+    on_macos do
+      destination = "/Library/Application Support/Mozilla/NativeMessagingHosts"
+    end
+
+    on_linux do
+      destination = "/usr/lib/mozilla/native-messaging-hosts"
+    end
+
+    <<~EOS
+      Before you can use the browser extension, you will need to manually link the app manifest
+      This is needed because Homebrew formulae cannot access directories outside Homebrew directory
+
+      #{Tty.magenta}#{Tty.bold}To link the manifest, run the following commands after the formula is installed:#{Tty.reset}
+      $ sudo mkdir -p #{destination}
+      $ sudo ln -sf "#{source}/#{filename}" "#{destination}/#{filename}"
+
+      #{Tty.red}You will not be able to use the extension until you link the manifest!#{Tty.reset}
+    EOS
+  end
+
+  test do
+    assert_match "firefoxpwa #{version}", shell_output("#{bin}/firefoxpwa --version")
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
Ok. I changed the test so it tries to launch a site, but it will always fail because that site does not exist and runtime is not installed. However, because of some changes that will come in the next project version, output will change to another error. I put a note documenting this in the test.